Multiple dimensions

sloede opened the pull request:
Closes #125, closes #90, closes #12, closes #144.

ranocha commented:
I've fixed some additional type instabilities and switch to `SVector{NDIMS, Float64}` instead of `Point`. Timing the second run of `Trixi2Vtk.run` for the initial output generated by `examples/3d/parameters_tgv.toml` resulted on my system in
```
 ────────────────────────────────────────────────────────────────────────────────────────
                                                 Time                   Allocations      
                                         ──────────────────────   ───────────────────────
            Tot / % measured:                 3.85s / 100%            1.59GiB / 100%     

 Section                         ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────────────────
 prepare VTK cells (node data)        1    2.62s  68.1%   2.62s   1.35GiB  85.0%  1.35GiB
 build VTK grid (node data)           1    803ms  20.9%   803ms    209MiB  12.9%   209MiB
 interpolate data                     1    259ms  6.73%   259ms   20.7MiB  1.27%  20.7MiB
 add data to VTK file                 1    129ms  3.35%   129ms   8.33MiB  0.51%  8.33MiB
   p                                  1   43.3ms  1.12%  43.3ms   37.1KiB  0.00%  37.1KiB
   v2                                 1   33.2ms  0.86%  33.2ms   8.05MiB  0.50%  8.05MiB
   v1                                 1   31.9ms  0.83%  31.9ms   37.1KiB  0.00%  37.1KiB
   rho                                1   11.7ms  0.30%  11.7ms   37.1KiB  0.00%  37.1KiB
   v3                                 1   7.78ms  0.20%  7.78ms   37.1KiB  0.00%  37.1KiB
   cell_ids                           1    307μs  0.01%   307μs   37.1KiB  0.00%  37.1KiB
   element_ids                        1    305μs  0.01%   305μs   49.3KiB  0.00%  49.3KiB
   levels                             1   92.2μs  0.00%  92.2μs   37.1KiB  0.00%  37.1KiB
 save VTK file                        2   28.2ms  0.73%  14.1ms   6.08KiB  0.00%  3.04KiB
 prepare VTK cells (cell data)        1   4.04ms  0.11%  4.04ms   2.93MiB  0.18%  2.93MiB
 read mesh                            1   1.74ms  0.05%  1.74ms    226KiB  0.01%   226KiB
 build VTK grid (cell data)           1   1.71ms  0.04%  1.71ms    473KiB  0.03%   473KiB
 read data                            1    879μs  0.02%   879μs   2.52MiB  0.15%  2.52MiB
 ────────────────────────────────────────────────────────────────────────────────────────
```

My last commit c76332a3d8be0c1a376b9b3645e1ccb781cee0ae changed that to
```
 ────────────────────────────────────────────────────────────────────────────────────────
                                                 Time                   Allocations      
                                         ──────────────────────   ───────────────────────
            Tot / % measured:                 1.65s / 100%             413MiB / 100%     

 Section                         ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────────────────
 build VTK grid (node data)           1    928ms  56.5%   928ms    209MiB  50.7%   209MiB
 prepare VTK cells (node data)        1    339ms  20.6%   339ms    175MiB  42.4%   175MiB
 interpolate data                     1    222ms  13.5%   222ms   20.7MiB  5.01%  20.7MiB
 add data to VTK file                 1    124ms  7.56%   124ms   4.32MiB  1.05%  4.32MiB
   p                                  1   42.0ms  2.56%  42.0ms   37.1KiB  0.01%  37.1KiB
   v2                                 1   34.1ms  2.07%  34.1ms   4.04MiB  0.98%  4.04MiB
   v1                                 1   28.8ms  1.75%  28.8ms   37.1KiB  0.01%  37.1KiB
   rho                                1   10.7ms  0.65%  10.7ms   37.1KiB  0.01%  37.1KiB
   v3                                 1   7.72ms  0.47%  7.72ms   37.1KiB  0.01%  37.1KiB
   cell_ids                           1    316μs  0.02%   316μs   37.1KiB  0.01%  37.1KiB
   element_ids                        1    284μs  0.02%   284μs   49.3KiB  0.01%  49.3KiB
   levels                             1   58.7μs  0.00%  58.7μs   37.1KiB  0.01%  37.1KiB
 save VTK file                        2   26.0ms  1.58%  13.0ms   6.08KiB  0.00%  3.04KiB
 read mesh                            1   1.75ms  0.11%  1.75ms    226KiB  0.05%   226KiB
 build VTK grid (cell data)           1   1.62ms  0.10%  1.62ms    473KiB  0.11%   473KiB
 read data                            1    887μs  0.05%   887μs   2.52MiB  0.61%  2.52MiB
 prepare VTK cells (cell data)        1    521μs  0.03%   521μs    693KiB  0.16%   693KiB
 ────────────────────────────────────────────────────────────────────────────────────────
```

sloede commented:
Wow, this is awesome! Now the bulk of the allocations happens inside `WriteVTK.jl`'s `vtk_grid` c'tor, which we cannot optimize easily ourselves. Great work!

ranocha commented:
Closes #125

sloede commented:
We hit 95% code coverage with coveralls.io. This would be good enough for me, I guess. Would everyone be OK with this? Also, shouldn't we disable Codecov? I think the faulty reporting is not really useful and the two different coverage percentages are confusing. Or is there another compelling reason to keep Codecov?

ranocha commented:
- I'm okay with the current coverage percentage.
- Codecov will post a report whereas coveralls is just reported as below. But I'm also okay with removing Codecov.

sloede commented:
Personally, I also like the way that Codecov.io presents its results much better. However, a report that decreases coverage is considered to be a failure and then it's confusing to find out which lines are really not covered and for which it's just a reporting snafu.

Thus, could you please remove it from Travis and from the README? If you just uncomment it, we can easily re-enable it in the future. And if I find the time, maybe I'll issue a report with Codecov.io.

ranocha commented:
I've disabled Codecov for Trixi.jl. To enable it again, we need to
- enable it in `.travis.yml`
- enable it on Codecov.io and/or the GitHuub integration

sloede commented:
I will try to further update the docs a little bit. My thinking goes as follows:
* In `README.md` I think we should document how a Trixi *user* would be able to install and use Trixi, i.e., via the Julia package mechanism.
* This should be kept as short as possible (while still being comprehensive) to highlight the ease of use.
* Further, there should also be a quick info on how to generate images from the solution files.

Then, I would refer to the docs for a description geared towards Trixi *developers*, where the installation & usage is explained by cloning the repo and using `julia --project=.`. Comments welcome!

ranocha commented:
Sounds good to me. That's exactly how I would like it to be. We should also do the same for Trixi2Vtk.jl and Trixi2Img.jl, I think.

So we would have something along the lines of

## Installation

If you have not yet installed Julia, please [follow the instructions for your operating system](https://julialang.org/downloads/platform/). Trixi works with Julia v1.5.

Trixi and related postprocessing tools are registered Julia packages. Hence, you can also install Trixi, [Trixi2Vtk](https://github.com/trixi-framework/Trixi2Vtk.jl), and [Trixi2Img](https://github.com/trixi-framework/Trixi2Img.jl) via
```julia
julia> import Pkg

julia> Pkg.add("Trixi"); Pkg.add("Trixi2Vtk"); Pkg.add("Trixi2Img")
```

Maybe we can have some sentence along the lines of 

If you do this and want to modify Trixi, you can run
```julia
julia> Pkg.dev("Trixi") # get a clone of the git repository, usually in ~/.julia/dev/Trixi
```

Then, we can describe a simple example, how to run one simulation and how to generate a plot of the solution.

sloede commented:
> Maybe we can have some sentence along the lines of [...]

To be honest, I wouldn't even include that in the main README. I wonder, who's the potential audience that would benefit from such a description:
* A novice Julia user will not understand the idea of a `dev` package without a more thorough explanation (at least I wouldn't have back then)
* An experienced Julia user knows how to do this anyways.
Thus I suggest to keep also this part of the explanation for the extended docs and really focus the main README on the essentials to get started. But we should definitely explicitly refer those who want to modify Trixi to the documentation such that they know where to look.

ranocha commented:
Okay, sounds good to me :+1:

sloede commented:
I've updated the README according to my ideas (but have not yet changed the docs). It would be great if you can have a look and review and/or modify it. Also, I changed the introductory paragraph to emphasize that it is not just a toy code for new users but also fast and easy to extend.

ranocha commented on README.md:
```diff
@@ -39,70 +41,48 @@ postprocessing.  Its features include:
+obtained by running `get_examples()`. Alternatively, you can also browse the
+[`examples/`](examples/) subdirectory. If you want to
+modify one of the parameter files to set up your own simulation, download it to
+your machine, edit the configuration, and pass the file path to `run(...)`.
```
`Trixi.run` instead of just `run`

sloede replied:
Fixed in f4e4e1c.

ranocha commented on README.md:
```diff
@@ -39,70 +41,48 @@ postprocessing.  Its features include:
+modify/extend Trixi's implementation, how to visualize output files etc. It
+also includes a section on our preferred development workflow and some tips for
+using Git. The latest documentation can be accessed either
 [online](https://trixi-framework.github.io/Trixi.jl/dev) or under [`docs/src`](docs/src).
```
Do we really want to point to the dev docs here instead of the stable docs? If a user has installed Trixi as described, they need the stable docs.

sloede replied:
Fixed in f4e4e1c.

ranocha commented:
Looks good to me. Here are two comments we might think about.
- Would it be good to tell users that the first run takes some time because of compilation while the second run will be fast? Otherwise, they might read that Trixi is performant at first, run the example one, and get disappointed.
- Shall we include a description how to visualize the output of the default example?

sloede commented:
> Looks good to me. Here are two comments we might think about.
> 
>     * Would it be good to tell users that the first run takes some time because of compilation while the second run will be fast? Otherwise, they might read that Trixi is performant at first, run the example one, and get disappointed.
> 
>     * Shall we include a description how to visualize the output of the default example?

Thanks, both are excellent suggestions. I've included them in f4e4e1c - if you could please review it once more?

ranocha commented on README.md:
````diff
@@ -64,26 +64,47 @@ In the Julia REPL, first load the package Trixi
+```
+and generate a contour plot of the results with
+```julia
+julia> Trixi2Img.convert("out/solution_000040.h5", grid_lines=true) # On Windows, use "out\\solution_000040.h5"
````
`joinpath("out", "solution_000040.h5")`

sloede replied:
TBH, I had it with `joinpath` at first but then decided to do it this way since the `joinpath` makes the command look much more complicated than it is. This in the spirit of giving the simplest explanation, I decided to hardcode it. But if you feel like this actually makes it more difficult to understand, I can also use the Julian approach.

ranocha commented:
Sure! I like your modified version :+1: I've just left one minor comment.

ranocha commented:
I think this will make it a bit more difficult for people using Windows, which seems to be the vast majority of students. Hence, I would prefer the approach that does not depend on the system - then, everybody can just copy & paste the instructions.
But I won't insist on this, so you can also keep it as it is.

sloede commented:
> then, everybody can just copy & paste the instructions.

No, you've convinced me. Fixed in 6ea0e49.

ranocha commented on README.md:
````diff
@@ -56,9 +57,22 @@ julia> Pkg.add("Trixi"); Pkg.add("Trixi2Vtk"); Pkg.add("Trixi2Img")
+```bash
+julia --project=.
+```
+Further details can be found in the [documentation](#documentation).
````
What about the possibility to
```julia
Pkg.add("Trixi")
```
at first (as a user) and
```julia
Pkg.dev("Trixi") # clones the repository, usually to ~/.julia/dev/Trixi on Linux
```
afterwards (as a developer)? At least that's the approach I prefer. Then, one doesn't have to start Julia with any special project flags or `instatiate` stuff.

ranocha replied:
To elaborate a bit: What's described in the current version is the approach for projects (in the sense of scientific projects like our repro repo). In contrast, what I've described above is what I consider the approach for packages (in the sense of libraries). I like to think of Trixi in the second way - let's provide building blocks to run simulations in a library instead of developing a code that has to be compiled and modified from scratch for each paper.

sloede replied:
OK, let's discuss this at the meeting later. While I generally agree with what you've said here, I personally like the "manual" approach better, as it allows to easily switch between different local Trixi folders without getting confused about which version I am currently using. But maybe I'm still doing it wrong ;-)

sloede replied:
Fixed in aa551e4 as per discussion.

sloede commented:
I've just updated the docs in such a way that I think they are ready to be published. Thus, my roadmap would be as follows:
* Resolve remaining issue above at the meeting this afternoon
* Merge `ndims` to `master`
* Once checks complete and docs look OK: bump version in `Project.toml` to `0.3.0`
* Register current master as new package
* Register current master for Trixi2Vtk & Trixi2Img as new package
* Update READMEs (badges)
* Make announcements on Discourse, Julia PDE repo, NumSim website
* Par-tey! :partying_face: 

If I'm missing something, please let me know at the meeting.

sloede commented:
@ranocha If this build goes through, could you maybe do a final check and then merge the PR to `master`? Then, it would be great if you could have a final look at the generated docs (to make sure I didn't do any stupid mistakes or typos) and then I think you could go ahead and proceed with steps 3-8 from https://github.com/trixi-framework/Trixi.jl/pull/129#issuecomment-683739642.

I will not be able to push anything until very late this evening, but I'll try to look into the comments from time to time if I notice any activity.
